### ISSUE TITLE
Where to add codes for saving checkpoints and loading them (train.py)?

### ISSUE THREAD

**Issue by KatieBelli** [tsurumeso/vocal-remover#122]
I researched about this but I couldn't figure out where to add those codes in the train.py file. After manual research on the internet I asked ChatGPT to explain and it wasn't helpful.

The suggestion for saving checkpoints from ChatGPT:

_"To save and load checkpoints in PyTorch, you typically need to add the code within the training loop. (...)"_

```
        # save checkpoint
        if i % 1000 == 999: # every 1000 mini-batches
            checkpoint = {
                'epoch': epoch,
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'loss': loss
            }
            torch.save(checkpoint, 'checkpoint.pth')
```
**Where does the training loop exactly start and does it really not matter where I add that code within the training loop?**

For loading checkpoints ("_To load a checkpoint and resume training from that point, you can add the following code at the beginning of your script_"):

```
# load checkpoint
checkpoint = torch.load('checkpoint.pth')
model.load_state_dict(checkpoint['model_state_dict'])
optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
epoch_start = checkpoint['epoch'] + 1
loss_start = checkpoint['loss']
```

**But where exactly at the beginning? To me, it sounds like I can put that code it into the first line but I know that's not the right way.**

Please be aware that I'm still learning and trying to understand how programming works.

**Comment by tsurumeso** [comment]
> Where does the training loop exactly start and does it really not matter where I add that code within the training loop?

Add below this line.
https://github.com/tsurumeso/vocal-remover/blob/3df56690088cf6e2e62049d0c0e745aa6bba5b44/train.py#L60

> But where exactly at the beginning? To me, it sounds like I can put that code it into the first line but I know that's not the right way.

Add below this line.
https://github.com/tsurumeso/vocal-remover/blob/3df56690088cf6e2e62049d0c0e745aa6bba5b44/train.py#L151

**Comment by KatieBelli** [comment]
Thanks. I added the codes for saving a checkpoint in the original train.py after line 60 in line 62:

    criterion = nn.CrossEntropyLoss()
    
    torch.save({
            'epoch': 3,
            'model_state_dict': model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'loss': criterion,
            }, 'checkpoints/model.pth')

It created a checkpoint pth-file [size: 169 MB (177.464.619 Bytes)]. I noticed that the size is bigger than the model size which always is 56,5 MB (59.317.925 Bytes). I hope this is a good sign. I don't know if I need  'epoch': , because I would let it train for 12 epochs on one day and 24 epochs the other day.

-------------------------------------------------------------------------------------

**New issue to fix:**
**How can I resume the training so that the resume always starts from the saved last best validation loss from the last saved training/checkpoint?**
My goal is that I can do several training sessions for one model so I can continue training that one model until the values are good enough (and so I can train that one model on different days in the future).

I resumed the checkpoint with those added lines but the validation loss in the first epoch was higher than the last epoch from the last training so I assume it either didn't save the values for training loss and validation loss or the code for resuming the training is not correct:

In my resume file for training (I copied the train.py file with the added code for saving checkpoints and renamed it "resume_train.py") I added those lines:

after line 10 "import torch.nn as nn":
`import torch.optim as optim` (=line 11 now)

after line 116 "p.add_argument('--patches', '-p', type=int, default=16)":
`lr = 0.001` (=line 117 now)

Now the code for resuming the training from the last checkpoint, after line 162, starting in line 164:
```
    model = model.to(device)

    optimizer = optim.Adam(model.parameters(), lr=lr, 
                       betas=(0.9, 0.999), 
                       eps=1e-8, weight_decay=0.0005)

    criterion = nn.CrossEntropyLoss()

    checkpoint = torch.load('C:/Users/.../Vocal Remover/vocal-remover/checkpoints/model.pth')
    model.load_state_dict(checkpoint['model_state_dict'])
    print('Previously trained model weights state_dict loaded...')
    optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
    print('Previously trained optimizer state_dict loaded...')
    epochs = checkpoint['epoch']
    criterion = checkpoint['loss']
    print('Trained model loss function loaded...')
    print(f"Previously trained for {epochs} number of epochs...")
```

-------------------------------------------------------------------------------------

PS: 
I copied those codes from https://debuggercafe.com/effective-model-saving-and-resuming-training-in-pytorch/
(I also asked ChatGPT for help and it did help with some problems I encountered but it doesn't help me now.)

**For saving checkpoints:**

```
# save model checkpoint
torch.save({
            'epoch': epochs,
            'model_state_dict': model.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'loss': criterion,
            }, '../outputs/model.pth')
```

**For resuming the training from the previous checkpoint:**

```
# load the trained model
model = model.CNN().to(device) # initilize the model
# initialize optimizer  before loading optimizer state_dict
optimizer = optim.Adam(model.parameters(), lr=lr, 
                       betas=(0.9, 0.999), 
                       eps=1e-8, weight_decay=0.0005)
```

```
# load the model checkpoint
checkpoint = torch.load('../outputs/model.pth')
# load model weights state_dict
model.load_state_dict(checkpoint['model_state_dict'])
print('Previously trained model weights state_dict loaded...')
# load trained optimizer state_dict
optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
print('Previously trained optimizer state_dict loaded...')
epochs = checkpoint['epoch']
# load the criterion
criterion = checkpoint['loss']
print('Trained model loss function loaded...')
print(f"Previously trained for {epochs} number of epochs...")
# train for more epochs
epochs = new_epochs
print(f"Train for {epochs} more epochs...")
```

-------------------------------------------------------------------------------------

It took me weeks to finally achieve that it saves a checkpoint because I'm still learning and am a beginner. I appreciate your help and I'm looking forward to your help/answer.
